contracts/libraries/SnappBaseCore.sol

```diff
@@ -0,0 +1,269 @@
+    event StateTransition(TransitionType transitionType, uint stateIndex, bytes32 stateHash, uint slot);
+    event SnappInitialization(bytes32 stateHash, uint8 maxTokens, uint16 maxAccounts);
+
+    enum TransitionType {
+        Deposit,
+        Withdraw
+    }
+
+    struct PendingBatch {
+        uint16 size;                   // Number of deposits in this batch
+        bytes32 shaHash;               // Rolling shaHash of all deposits
+        uint creationTimestamp;        // Timestamp of batch creation
+        uint appliedAccountStateIndex; // accountState index when batch applied (for rollback), 0 implies not applied.
+    }
+
+    struct ClaimableWithdrawState {
+        bytes32 merkleRoot;                      // Merkle root of claimable withdraws in this block
+        bool[WITHDRAW_BATCH_SIZE] claimedBitmap; // Bitmap signalling which withdraws have been claimed
+        uint appliedAccountStateIndex;           // AccountState when this state was created (for rollback)
+    }
+
+    struct Data {
+        bytes32[] stateRoots;
+        IdToAddressBiMap.Data registeredAccounts;
+        IdToAddressBiMap.Data registeredTokens;
+        uint8 numTokens;
+        uint depositIndex;
+        mapping (uint => PendingBatch) deposits;
+        uint withdrawIndex;
+        mapping (uint => PendingBatch) pendingWithdraws;
+        mapping (uint => ClaimableWithdrawState) claimableWithdraws;
+    }
+
+    function init(Data storage data) public {
+        // The initial state should be Pederson hash of an empty balance tree
+        bytes32 stateInit = bytes32(0);
+        data.stateRoots.push(stateInit);
+        data.depositIndex = MAX_UINT;
+        data.withdrawIndex = MAX_UINT;
+
+        emit SnappInitialization(stateInit, MAX_TOKENS, MAX_ACCOUNT_ID);
+    }
+
+    /**
+     * Public View Methods
+     */
+    function stateIndex(Data storage data) public view returns (uint) {
+        return data.stateRoots.length - 1;
+    }
+
+    function publicKeyToAccountMap(Data storage data, address addr) public view returns (uint16) {
+        return IdToAddressBiMap.getId(data.registeredAccounts, addr);
+    }
+
+    function accountToPublicKeyMap(Data storage data, uint16 id) public view returns (address) {
+        return IdToAddressBiMap.getAddressAt(data.registeredAccounts, id);
+    }
+
+    function tokenIdToAddressMap(Data storage data, uint16 id) public view returns (address) {
+        return IdToAddressBiMap.getAddressAt(data.registeredTokens, id);
+    }
+
+    /**
+     * General Snapp Functionality
+     */
+    function openAccount(Data storage data, uint16 accountId) public {
+        require(accountId < MAX_ACCOUNT_ID, "Account index exceeds max");
+        require(
+            IdToAddressBiMap.insert(data.registeredAccounts, accountId, msg.sender),
+            "Address occupies slot or requested slot already taken"
+        );
+    }
+
+    function addToken(Data storage data, address _tokenAddress) public {
```

Comment:
Having this big struct data implements everything very elegantly.

I am wondering, whether we have to buy this at a cost. Do we have to "pay" gas for calling the function with the whole struct although only two variables are used?